usr/src/cmd/mdb/aarch64/kmdb/kaif.c

```diff
@@ -0,0 +1,663 @@
+		    wp->wp_size);
+		return (set_errno(EINVAL));
+	}
+
+	return (0);
+}
+
+static int
+kaif_wapt_reserve(kmdb_wapt_t *wp)
+{
+	size_t wrps = 0;
+	int id;
+
+	wrps = kmdb_kdi_num_wapts();
+
+	for (id = 0; id < wrps; id++) {
+		if (!BT_TEST(&kaif_waptmap, id)) {
+			/* found one */
+			BT_SET(&kaif_waptmap, id);
+			wp->wp_priv = (void *)(uintptr_t)id;
+			return (0);
+		}
+	}
+
+	return (set_errno(EMDB_WPTOOMANY));
+}
+
+static void
+kaif_wapt_release(kmdb_wapt_t *wp)
+{
+	int id = KAIF_WPPRIV2ID(wp);
+
+	ASSERT(BT_TEST(&kaif_waptmap, id));
+	BT_CLEAR(&kaif_waptmap, id);
+}
+
+static void
+kaif_wapt_arm(kmdb_wapt_t *wp)
+{
+	kdi_waptreg_t waptreg = {0};
+	uint_t hwid = KAIF_WPPRIV2ID(wp);
+	uint_t rw = 0;
+
+	ASSERT(BT_TEST(&kaif_waptmap, hwid));
+
+	if (wp->wp_wflags & MDB_TGT_WA_R)
+		rw |= WCR_LSC_LOAD;
+	if (wp->wp_wflags & MDB_TGT_WA_W)
+		rw |= WCR_LSC_STORE;
+
+	/*
+	 * byte-address-select, a mask of the bytes following wp_addr which
+	 * are watched.  Not necessarily starting at 0.
+	 */
+	uint8_t bas = 0xff;
+	uintptr_t addr = P2ALIGN(wp->wp_addr, 8);
+	ptrdiff_t addr_phase = P2PHASE(wp->wp_addr, 8);
+	uint8_t mask = 0x0;
+
+	/*
+	 * double-word aligned address, 8-byte byte-address-select
+	 * word aligned address, 4-byte byte-address select
+	 */
+	if ((IS_P2ALIGNED(wp->wp_addr, 8) && wp->wp_size <= 8) ||
+	    (IS_P2ALIGNED(wp->wp_addr, 4) && wp->wp_size <= 4)) {
+		bas = (1 - wp->wp_size) << addr_phase;
+	} else {
+		mask = (1 << wp->wp_size) - 1;
+	}
+
+	ASSERT((mask == 0x0) || (bas == 0xff));
+
+	waptreg.kw_hwid = hwid;
+	waptreg.kw_addr = addr;
+	waptreg.kw_ctl = WCR_MASK(mask) | WCR_BAS(bas) | WCR_LSC(rw) |
+	    WCR_PAC(1) | WCR_ENABLE;
+
+	kmdb_kdi_update_waptreg(&waptreg);
+
+}
+
+static void
+kaif_wapt_disarm(kmdb_wapt_t *wp)
+{
+	kdi_waptreg_t new = {0};
+	int hwid = KAIF_WPPRIV2ID(wp);
+
+
+	ASSERT(BT_TEST(&kaif_waptmap, hwid));
+
+	kmdb_kdi_read_waptreg(hwid, &new);
+
+	new.kw_ctl &= ~WCR_ENABLE;
+
+	kmdb_kdi_update_waptreg(&new);
+}
+
+static int
+kaif_wapt_match(kmdb_wapt_t *wp)
+{
+	uint64_t far = 0;
+	int hwid = KAIF_WPPRIV2ID(wp);
+	int i;
+
+	ASSERT(BT_TEST(&kaif_waptmap, hwid));
+
+	kmdb_dpi_get_register("far", &far);
+
+	for (i = 0; i < kmdb_kdi_num_wapts(); i++) {
+		kdi_waptreg_t kw;
+
+		kmdb_kdi_read_waptreg(i, &kw);
+
+		if ((far >= wp->wp_addr) &&
+		    (far < (wp->wp_addr + wp->wp_size))) {
+			return (1);
+		}
+	}
+
+	return (0);
+}
+
+static int
+kaif_step(void)
+{
+	kreg_t spsr;
+
+	/*
+	 * set single-step, disable interrupts, and clear PSTATE.D when we
+	 * resume.
+	 */
+	kmdb_dpi_get_register("spsr", &spsr);
+	kmdb_dpi_set_register("spsr", (spsr | PSR_SS | PSR_I) & ~PSR_D);
+	write_mdscr_el1(read_mdscr_el1() | MDSCR_SS);
+
+	kmdb_dpi_resume_master(); /* Run the next instruction and come back */
+
+	/* restore the single step setting and mask settings */
+	kmdb_dpi_set_register("spsr", spsr);
+	write_mdscr_el1(read_mdscr_el1() & ~MDSCR_SS);
+
+	return (0);
+}
+
+/*ARGSUSED*/
```

Comment:
Delete the comment? All arguments are used.

Reply:
I'll get this, and clean up the others `ARGSUSED` in files I touched to use `__unused` unless they're really ignoring _every_ argument